toolchain/lex/lex.cpp

```diff
@@ -997,46 +991,22 @@ auto Lexer::LexOpeningSymbolToken(llvm::StringRef source_text, TokenKind kind,
 
 auto Lexer::LexClosingSymbolToken(llvm::StringRef source_text, TokenKind kind,
                                   ssize_t& position) -> LexResult {
-  auto unmatched_error = [&] {
-    CARBON_DIAGNOSTIC(UnmatchedClosing, Error,
-                      "Closing symbol without a corresponding opening symbol.");
-    emitter_.Emit(source_text.begin() + position, UnmatchedClosing);
-    TokenIndex token = buffer_.AddToken({.kind = TokenKind::Error,
-                                         .token_line = current_line(),
-                                         .column = ComputeColumn(position),
-                                         .error_length = 1});
-    ++position;
-    return token;
-  };
+  TokenIndex token = LexOneCharSymbolToken(source_text, kind, position);
 
-  // If we have no open groups, this is an error.
   if (LLVM_UNLIKELY(open_groups_.empty())) {
-    return unmatched_error();
+    has_mismatched_brackets_ = true;
+    return token;
   }
 
-  TokenIndex opening_token = open_groups_.back();
-  // Close any invalid open groups first.
-  if (LLVM_UNLIKELY(buffer_.GetTokenInfo(opening_token).kind !=
+  if (LLVM_UNLIKELY(buffer_.GetTokenInfo(open_groups_.back()).kind !=
                     kind.opening_symbol())) {
-    CloseInvalidOpenGroups(kind, position);
-    // This may exhaust the open groups so re-check and re-error if needed.
-    if (open_groups_.empty()) {
-      return unmatched_error();
-    }
-    opening_token = open_groups_.back();
-    CARBON_DCHECK(buffer_.GetTokenInfo(opening_token).kind ==
-                  kind.opening_symbol());
+    has_mismatched_brackets_ = true;
+    return token;
   }
-  open_groups_.pop_back();
 
-  // Now that the groups are all matched up, lex the actual token.
-  TokenIndex token = LexOneCharSymbolToken(source_text, kind, position);
-
-  // Note that it is important to get fresh token infos here as lexing the
-  // open token would invalidate any pointers.
+  TokenIndex opening_token = open_groups_.pop_back_val();
   buffer_.GetTokenInfo(opening_token).closing_token = token;
```

Comment:
Can do this once above the `if` now and keep the token info alive for the duration.

Similarly, it's probably best to get the token info for the just-lexed token immediately after lexing it.

Reply:
Done.